### DISCUSSION TITLE
Is this package pure ESM? getting "require() of ES Module not supported." error

### DISCUSSION THREAD

**Discussion by stonesthatwhisper** [igordanchenko/yet-another-react-lightbox#63] [Q&A] [answered]
`package.json` in my project does not have `"type": "module"`, and I am seeing this error. Is this package "pure ESM"? Do I have to [convert my project](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) to pure ESM to use it? Is this (be a pure ESM project) a prerequisite for using this package? 

```
Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/me/project/node_modules/yet-another-react-lightbox/dist/index.js from /Users/me/project/.next/server/pages/project.js not supported.
Instead change the require of index.js in /Users/me/project/.next/server/pages/project.js to a dynamic import() which is available in all CommonJS modules.
    at Object.require.extensions.<computed> [as .js] (/Users/me/project/node_modules/ts-node/dist/index.js:851:20)
    at /Users/me/project/.next/server/pages/project.js:711:72
    at runMicrotasks (<anonymous>) {
```

**Comment by igordanchenko** [comment, marked as answer]
> Is this package "pure ESM"?

Yes, this package is published as ESM-only.

> Do I have to [convert my project](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) to pure ESM to use it?

Absolutely not. Since you are using a bundler, you don't have to set `"type": "module"` in your project's `package.json` in order to use this library. All you need to do is to `import` this library instead of trying to `require` it. By the way, is there any particular reason you are using `require()` syntax in a Next.js project instead of `import`?

**Reply by stonesthatwhisper** [reply]
Thanks for the explanation. I really appreciate it.

I am using `import` in the Next.js project, however, tsconfig.json has
`{ "compilerOptions": {
    "module": "commonjs"
}
}`

Maybe this is the reason the compiled code has `require`, and it doesn't work well with this package?

**Reply by igordanchenko** [reply]
Ah, I haven't thought of `tsconfig.json`, but it doesn't look like `"module": "commonjs"` is the only culprit. I tried to reproduce this error on my end, but no luck so far. What are your Next.js, node and TypeScript versions? Are you using any plugins in your `next.config.js`? Any webpack customizations?

**Reply by stonesthatwhisper** [reply]
If you are not seeing this error with `"module": "commonjs"` that's good news. However once I change it to anything else like "es5" I got:
`(node:54502) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.`
It looks like a no-go, and stick to "commonjs" is the only way.

However, when I get the error
`[ERR_REQUIRE_ESM]: require() of ES Module /Users/me/project/node_modules/yet-another-react-lightbox/dist/index.js from /Users/me/project/.next/server/pages/project.js`
I went to check the `project.js` file and did see the `require()` call in there. I thought once you set  `"module": "commonjs"` you should be able to replicate. Is my NextJS version too low?

These are the versions:
```
next@12:
  version "12.3.4"

typescript@4.3.2:
  version "4.3.2"

node@14
or node@16
```

next.config.js. (I am using next-plugin-antd-less. not sure if that is the culprit.) No other webpack customization I know of.
```
require('dotenv').config();
const withAntdLess = require('next-plugin-antd-less');
module.exports = withAntdLess({
  env: {...},
  nextjs: {
    localIdentNameFollowDev: true, // default false, for easy to debug on PROD mode
  },
  webpack(config) {
    return config;
  },
});
```

I find a `.babelrc`. In the past I just add stuff into it per Next's recommendation.
```
{
  "presets": [
    [
      "next/babel",
      {
        "class-properties": { "loose": true }
      }
    ]
  ],
  "plugins": [
    [ "import", { "libraryName": "antd", "style": true } ],
    [ "module-resolver", { "alias": { "@": "./src" } }],
    ["@babel/plugin-proposal-private-property-in-object", { "loose": true }],
    ["@babel/plugin-proposal-private-methods", { "loose": true }]
  ]
}
```

**Reply by igordanchenko** [reply]
Thank you for providing the additional details. Unfortunately I'm still unable to reproduce. I tried node v14.21.2, Next v12.3.4, TypeScript v4.3.2, and your `next.config.js` with `next-plugin-antd-less` plugin, and everything works just fine.

By the way, the `"module"` prop in `tsconfig.json` should normally be set to `"esnext"`. Here is the default `tsconfig.json` created by `create-next-app`:

```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
  "exclude": ["node_modules"]
}
```

And, of course, you do not need to set "type": "module" in the package.json.

I'll be happy to take a look if you can provide a complete isolated demo of the project. Or you can start by creating an empty project with `create-next-app`, and incrementally add features from your project to see when it breaks. 

Oh, and I'd also make sure the node version is one of the recent ones in the corresponding release branch (v14/v16/v18), and not a pre-LTS version.

I hope this helps.

**Reply by igordanchenko** [reply]
I just saw you updated your last comment with `.babelrc` info. Can you provide your babel and babel plugin versions?

**Reply by stonesthatwhisper** [reply]
> Can you provide your babel and babel plugin versions?

I am not sure how to find babel version. Here are the related lines in package.json. I really appreciate you troubleshooting this. 
```
devDependencies: {
  ...
  "@babel/plugin-proposal-private-methods": "^7.18.6",
    "@babel/plugin-proposal-private-property-in-object": "^7.20.5",
   "babel-plugin-import": "^1.13.5",
    "babel-plugin-module-resolver": "^4.1.0",
...
}
```

I would really like to set `"module": "esnext"` but am getting `(node:54502) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.` right away.

**Reply by igordanchenko** [reply]
Alright, I tried including babel in my test, but it still builds and runs just fine. We are missing something here. Can you show your complete `tsconfig.json` and `package.json`? (please be sure to remove all proprietary or sensitive items) 

Are you using `next/dynamic` to load the lightbox? Can you show your page with the lightbox?

I'm not aware of a reason why you shouldn't be able to run Next.js project with `"module": "esnext"` in your `tsconfig.json`. Did you have to change "module" to "commonjs" before you tried using YARL? This doesn't sound right...

**Reply by igordanchenko** [reply]
@stonesthatwhisper 

I'm following up to see if you were able to get to the bottom of this issue. I'm really curious to figure out what exactly in your project configuration is affecting your dev experience. I'll be more than happy to take a closer look if you could share a trimmed-down version of your project (remove all pages and components except the one with the lightbox). Please let me know.

**Reply by stonesthatwhisper** [reply]
@igordanchenko Thank you for offering help. It's been a major pain for me to move to esnext. When I actually set module to "esnext",  "yarn build" can complete successfully however  "yarn start" would fail with "SyntaxError: Cannot use import statement outside a module". I'll try to get a trimmed-down version of the project, but that'll take some time.

Nevertheless, I found out my project works with YARL if I build and run it ("yarn build; yarn start") with `"module": "commonjs"` in `tsconfig.json`.

The error I reported in the initial post only happens in dev mode with "yarn dev" that runs `ts-node --ignore .next -- server/app.ts`.

Here is my complete `tsconfig.json`:
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "target": "es2017",
    "module": "commonjs",   // "ES2015": yarn build works, but yarn run fails.
    "jsx": "preserve",
    "allowJs": true,
    "alwaysStrict": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "removeComments": true,
    "preserveConstEnums": true,
    "sourceMap": false,
    "skipLibCheck": true,
    "baseUrl": ".",
    "typeRoots": ["./node_modules/@types"],
    "lib": ["dom", "es2017"],
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "isolatedModules": false,
    "noEmit": false,
    "outDir": "production-server/"
  },
  "include": [
    "next-env.d.ts",
    "./**/*.tsx",
    "./**/*.ts",
    "./**/*.js",
    "./server/**/*.ts"
  ],
  "exclude": [
    "dist",
    "production-server",
    ".next",
    "out",
    "next.config.js",
    "node_modules"
  ]
}
```

**Reply by igordanchenko** [reply]
> "yarn dev" that runs `ts-node --ignore .next -- server/app.ts`

That's the critical piece of information that was missing. A typical Next.js project uses `next dev` as its "dev" script.

```
"scripts": {
  "dev": "next dev",
  "build": "next build",
  "start": "next start",
  "lint": "next lint"
}
```

But since you are using `ts-node` to run your app in dev mode (I'm honestly not sure how it actually works in terms of substituting everything that Next.js does as a framework), it kind of explains the commonjs / esm issues that you are seeing.

> I'll try to get a trimmed-down version of the project, but that'll take some time.

That's pretty much the only option for me to assist you at this point since you are running a non-standard Next.js project.

**Reply by stonesthatwhisper** [reply]
Thanks @igordanchenko. I'm trying to switch to 'next dev' for our dev environment.
